[OID4VCI] VC of type oid4vc_natural_person has invalid id value

tdiesler opened the pull request:
closes #43854

This PR makes a number of interrelated changes in connection with VC Subject Id.

In the order of file changes, we have ...

* IssuedAt (iat) becomes a disclosed subject claim (not an SD top level claim by spec)
* Improved error messages on invalid disclosures (i.e. name the invalid claim)
* Add default user attributes for --feature=oid4vc-vci (specifically, it adds the User's DID)
* SubjectIdMapper uses the User's DID instead of a hard coded value or a random uuid
* SdJwtCredentialBuilder adds expiry date (exp) as top level claim (by spec)
* SdJwtCredentialBuilder adds subject id (sub) as disclosed subject claim

tdiesler committed:
[OID4VCI] VC of type oid4vc_natural_person has invalid id value

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

Captain-P-Goldfish commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCSubjectIdMapper.java:
```diff
@@ -42,61 +44,78 @@ public class OID4VCSubjectIdMapper extends OID4VCMapper {
     public static final String MAPPER_ID = "oid4vc-subject-id-mapper";
 
+    public static final String CLAIM_NAME_ID = "id";
+    public static final String USER_ATTRIBUTE_DID = "did";
```
should be moved to `OID4VCConstants`

Captain-P-Goldfish reviewed (changes requested):
not finished with the review yet. But I am once more out of time :-)

tdiesler commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/credentialbuilder/SdJwtCredentialBuilder.java:
```diff
@@ -47,48 +51,53 @@ public SdJwtCredentialBody buildCredentialBody(
+
+        // Put inner claims into the disclosure spec, except the one to be kept visible
+        var disclosureSpecBuilder = DisclosureSpec.builder();
+        var outerClaims = credentialBuildConfig.getSdJwtVisibleClaims();
```
done

tdiesler committed:
-- address reviewer feedback

-- fix duplicate salt in DisclosureRedListTest
-- move constants to OID4VCConstants
-- remove var style declarations

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

tdiesler committed:
-- fix OID4VCTimeNormalizationSdJwtTest as 'iat' is now an inner claim

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

Captain-P-Goldfish commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/credentialbuilder/SdJwtCredentialBuilder.java:
```diff
@@ -47,47 +55,52 @@ public String getSupportedFormat() {
-                    } else {
-                        disclosureSpecBuilder.withUndisclosedClaim(entry.getKey(), SdJwtUtils.randomSalt());
-                    }
+        List<String> outerClaims = credentialBuildConfig.getSdJwtVisibleClaims();
```
could you rename this variable to `visibleClaims`?

Captain-P-Goldfish commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCSubjectIdMapper.java:
```diff
@@ -49,54 +54,68 @@ public class OID4VCSubjectIdMapper extends OID4VCMapper {
+                .orElse(null);
+        if (userAttr != null) {
+            claims.put(propertyName, userAttr);
+        }
```
You are reading the subject hardcoded by the `did`-key, and you ignore the configured value under `ID Property Name`.
I have no objections to set the subject to null if no value is present, but I doubt that we should use the `did`-value as default, as did is a specific use-case.

I would say the default should stay with `id` and the attribute-name should be configurable:

```java
UserModel userModel = userSessionModel.getUser();
        List<String> attributePath = getMetadataAttributePath();
        String propertyName = attributePath.get(attributePath.size() - 1);
        String userAttributeName = mapperModel.getConfig().get(OID4VCMapper.USER_ATTRIBUTE_KEY);
        String userAttr = KeycloakModelUtils.resolveAttribute(userModel, userAttributeName ,false).stream()
                .filter(Objects::nonNull)
                .findFirst()
                .orElseGet(() -> {
                   return "id".equals(userAttributeName) ? userModel.getId() : null;
                });
```

tdiesler replied:
ok, will do ...

Captain-P-Goldfish replied:
the userAttribute cannot be set to "id".We have one use-case at least where this will be necessary.

Captain-P-Goldfish replied:
yes, that is much better than my proposal! :-)
thx.

IngridPuppet commented on server-spi-private/src/main/java/org/keycloak/models/oid4vci/CredentialScopeModel.java:
```diff
@@ -42,8 +42,10 @@
-    public static final String SD_JWT_VISIBLE_CLAIMS_DEFAULT = "id,iat,nbf,exp,jti";
+    // Also see DisclosureRedList#redList
+    // https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-11.html#section-3.2.2.2
+    public static final String SD_JWT_VISIBLE_CLAIMS_DEFAULT = "iss,nbf,exp,cnf,vct,status";
```
Guard against correlation by default?

Captain-P-Goldfish replied:
Yes, I’m aware of this section.  
However, from a use-case perspective, I would argue that `id`, `iat`, and especially `jti` are disclosed only in very rare scenarios. In typical usage, these fields should be visible, and since their visibility remains configurable, I believe the default should prioritize convenience for most users.

IngridPuppet reviewed (commented):
Hello @tdiesler and @Captain-P-Goldfish - I checked the PR and left some comments. Thank you, @tdiesler, for contributing the PR. Please could you check my comments?

tdiesler committed:
-- minor changes in response to reviews - thanks for doing this

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

tdiesler committed:
-- preserve configurable user attr - 'did' by default

-- use verifiableCredential.getExpirationDate()

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

tdiesler committed:
-- allow userModel.id to be used as subject id

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

tdiesler committed:
-- verify correct did -> sub id mapping

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

tdiesler committed:
-- review sd-jwt jti, id, sub

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

tdiesler committed:
-- align list of visible claims in js and doc

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

tdiesler committed:
-- align OID4VCSubjectIdMapper display type with Admin UI tests

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

Captain-P-Goldfish commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCGeneratedIdMapper.java:
```diff
@@ -65,29 +68,33 @@ protected List<ProviderConfigProperty> getIndividualConfigProperties() {
+        String propertyName = attributePath.get(attributePath.size() - 1);
+        if (CLAIM_NAME_VC_ID.equals(propertyName)) {
+            verifiableCredential.setId(vcId);
+        } else {
```
I actually thought we should set the id of the user instead of generating a random UUID in this case?

Captain-P-Goldfish commented on core/src/main/java/org/keycloak/representations/userprofile/config/UPConfig.java:
```diff
@@ -111,6 +111,16 @@ public UPAttribute getAttribute(String name) {
+            }
+        }
+        return false;
+    }
```
method is unused

Captain-P-Goldfish commented on core/src/main/java/org/keycloak/sdjwt/ClaimVerifier.java:
```diff
@@ -495,23 +489,17 @@ public Builder withExpCheck() {
-                        && ((ClaimCheck) verifier).getClaimName().equalsIgnoreCase(OID4VCConstants.CLAIM_NAME_EXP));
-            });
+            contentVerifiers.removeIf(verifier ->
+                    isTypeOrNamedCheck(verifier, ClaimVerifier.ExpCheck.class, OID4VCConstants.CLAIM_NAME_EXP));
```
you have remove the fallback checks in case that custom-verifiers were added. Please readd these checks for the methods

tdiesler committed:
-- remove urn:uuid prefix from jti

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

tdiesler committed:
-- clarify usage of subject id

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

tdiesler committed:
-- use claim name 'id' for subject id, keep credential id with vc

-- always generate a vc.id, also when oid4vc-generated-id-mapper is not installed

Signed-off-by: Thomas Diesler <tdiesler@ibm.com>

Captain-P-Goldfish commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/credentialbuilder/JwtCredentialBuilder.java:
```diff
@@ -90,12 +93,15 @@ public JwtCredentialBody buildCredentialBody(
+        CredentialSubject subject = verifiableCredential.getCredentialSubject();
+        Optional.ofNullable(subject.getClaims().remove(CLAIM_NAME_SUBJECT_ID)).ifPresent( id -> {
+            subject.getClaims().put(ID_CLAIM_KEY, id);
+        });
```
We have a usecases where we want to use the ids from remote providers. We have IdPs configured with the transient-users feature. It is not possible to set a specific ID for the transient user so we need to use the user-attributes. We want also to preserve the returned IDs from the IdP for the vc.id and the subject.id. In that case this part would be a problem.

tdiesler replied:
Understood. I think I found a way to preserve the mapping to the subject 'id' claim from an arbitrary user attribute while at the same time removing the conflict with vc.id in the claim model.

In the SdJwtCredentialBuilder we now do this ...

```
        // Map subject id => sub
        Optional.ofNullable(claims.remove(CLAIM_NAME_SUBJECT_ID)).ifPresent(it ->
                claims.put(CLAIM_NAME_SUB, it)
        );

        // Add credential id
        Optional.ofNullable(vcId).ifPresent(it ->
                claims.put(CLAIM_NAME_VC_ID, it)
        );
```

The other CredentialBuilders can use the subject 'id' claim without additional mapping hacks. 
Would this work for you?
